uplink/config.go

```diff
@@ -40,10 +43,58 @@ type RSConfig struct {
 // EncryptionConfig is a configuration struct that keeps details about
 // encrypting segments
 type EncryptionConfig struct {
-	Key       string      `help:"root key for encrypting the data"`
-	BlockSize memory.Size `help:"size (in bytes) of encrypted blocks" default:"1KiB"`
-	DataType  int         `help:"Type of encryption to use for content and metadata (1=AES-GCM, 2=SecretBox)" default:"1"`
-	PathType  int         `help:"Type of encryption to use for paths (0=Unencrypted, 1=AES-GCM, 2=SecretBox)" default:"1"`
+	// TODO: WIP#if/v3-1541 read TODO in Key method why the following field cannot
+	// be added
+	// RootKey     *storj.Key  // The root key for encrypting the data and read it from KeyFilepath, see Key method.
+	KeyFilepath string      `help:"the path to the file which contains the root key for encrypting the data"`
+	BlockSize   memory.Size `help:"size (in bytes) of encrypted blocks" default:"1KiB"`
+	DataType    int         `help:"Type of encryption to use for content and metadata (1=AES-GCM, 2=SecretBox)" default:"1"`
+	PathType    int         `help:"Type of encryption to use for paths (0=Unencrypted, 1=AES-GCM, 2=SecretBox)" default:"1"`
+}
+
+// Key returns the root key for encrypting the data which is stored in the path
+// indicated by KeyFilepath field.
+//
+// It returns an error if:
+//
+// * The file doesn't exist.
+// * The file doesn't have exactly 0400 permissions.
+// * There is an I/O error.
+//
+// TODO: WIP#if/v3-1541 refactor cache the key after the firs read. The problem
+// is with pkg/cfgstruct which is inflexible on accepting unexported keys nor
+// several different types as a type of storj.Key, [32]byte, etc.
+func (encCfg *EncryptionConfig) Key() (storj.Key, error) {
+	if encCfg.KeyFilepath == "" {
+		return storj.Key{}, nil
```

Comment:
This seems really dangerous, return a specific error `ErrKeyNotSpecified` or similar.

Reply:
My understanding is that currently it was possible to have an empty key and the configuration file contained `enc.key: ""` and this try to have the same behavior.

Not in my likes of not forcing to have a key being hones.

Reply:
The issue I can see is that someone will use:

```
key, err := enc.Key()
if err == nil { return err }

// start using the key as if it were valid
```

Reply:
I totally agree, but I guess that if we try to force a key there will be several things that can be broken.

For example I see that uplink setup ask for a passphrase, it can be empty so I left the keyfile path empty which means that there is not passphrase.

In order of not breaking the current behavior, I see that rather than doing this, we could do one of the following options:

* Return a specific error (e.g. `var NoKeyErr`) and the caller can identify it an assume the risk if it wants.
* Store a file with a zero key, so it's always expected to find a keyfile.

Do you have any other idea?

My understanding from the initial team green meeting of this sprint is to have an initial implementation and improve it from this point, so if the solution requires too many changes then it shouldn't classify on such agreement, nonetheless I don't have any issue on extending this work if we believe that's better for the sake of something like security.